common/XenomaiLock.cpp

```diff
@@ -0,0 +1,194 @@
+            initialize_xenomai();
+            if (int ret = __wrap_pthread_mutex_init(&mutex, NULL)) {
+                fprintf(stderr, "Error: unable to initialize mutex : (%d) %s\n", ret, strerror(-ret));
+                return;
+            }
+        }
+    }
+    enabled = true;
+}
+
+XenomaiMutex::~XenomaiMutex() {
+    xprintf("Destroy mutex %p\n", &mutex);
+    if (enabled)
+        __wrap_pthread_mutex_destroy(&mutex);
+}
+
+// a helper function to try
+// - call func()
+// - if it fails, try to turn the current thread into a cobalt thread and call func() again
+// - if it fails again, just fail
+// - return true if it succeeds, or false if it fails
+// id and name are just for debugging purposes, while enabled is there because it saves duplicating some lines
+template <typename T> static int try_or_retry(std::function<int()> func, T* id, const char* name, bool enabled) {
+    xprintf("tid: %d ", get_tid());
+    int ret;
+    if (!enabled) {
+        xfprintf(stderr, "%s %d disabled %p\n", name, id);
+        return false;
+    }
+    xprintf("%s %p\n", name, id);
+    if (EPERM != (ret = func())) // "success" (or at least meaningful failure)
+        return ret;
+    // if we got EPERM, we are not a Xenomai thread
+    if (turn_into_cobalt_thread()) {
+        xfprintf(stderr, "%s %p could not turn into cobalt\n", name, id);
+        return EPERM;
+    }
+    if (EPERM == (ret = func()))
+        xfprintf(stderr, "%s %p failed after having turned into cobalt: %d\n", name, id, ret);
+    return ret;
+}
+
+// condition resource_deadlock_would_occur instead of deadlocking. https://en.cppreference.com/w/cpp/thread/mutex/lock
+bool XenomaiMutex::try_lock(bool recurred) {
+    const char name[] = "try_lock";
+    return 0 == try_or_retry([this]() { return __wrap_pthread_mutex_trylock(&this->mutex); }, &mutex, name, enabled);
+    // TODO: An implementation that can detect the invalid usage is encouraged to throw a std::system_error with error
+    // condition resource_deadlock_would_occur instead of deadlocking.
+}
+
+void XenomaiMutex::lock(bool recurred) {
+    const char name[] = "lock";
+    try_or_retry([this]() { return __wrap_pthread_mutex_lock(&this->mutex); }, &mutex, name, enabled);
+}
+
+void XenomaiMutex::unlock(bool recurred) {
+    const char name[] = "unlock";
+    try_or_retry([this]() { return __wrap_pthread_mutex_unlock(&mutex); }, &mutex, name, enabled);
+}
+
+XenomaiConditionVariable::XenomaiConditionVariable() {
+    xprintf("Construct CondictionVariable\n");
+    if (int ret = __wrap_pthread_cond_init(&cond, NULL)) {
+        if (EPERM == ret) {
+            xprintf("mutex init returned EPERM\n");
+            initialize_xenomai();
+            if (int ret = __wrap_pthread_cond_init(&cond, NULL)) {
+                fprintf(stderr, "Error: unable to create condition variable : (%d) %s\n", ret, strerror(ret));
+                return;
+            }
+        }
+        return;
```

Comment:
the constructors of XenomaiConditionVariable and XenomaiMutex seem to do basically the same thing, but have slightly different structures because of this `return`. is that intentional? it would seem to me that we want to set `enabled = true` if either call to `__wrap_pthread_[cond,mutex]_init` succeeds, and otherwise leave it `false`.

Reply:
that's correct, see 3d87c79cb.